pvlib/horizon.py

```diff
@@ -0,0 +1,651 @@
+                right = np.array([lat_grid[i, j+1],
+                                  lon_grid[i, j+1],
+                                  elev_grid[i, j+1]])
+                bottom = np.array([lat_grid[i+1, j],
+                                   lon_grid[i+1, j],
+                                   elev_grid[i+1, j]])
+                samples = uniformly_sample_triangle(center,
+                                                    bottom,
+                                                    right,
+                                                    samples_per_triangle)
+                all_samples = np.vstack([all_samples, samples])
+    return np.array(all_samples)
+
+
+def _sample_using_interpolator(lat_grid, lon_grid, elev_grid, num_samples):
+    """
+    Creates a "grid" using polar coordinates and uses the scipy's grid
+    interpolator to estimate elevation values at each point on the polar grid
+    from the input (rectangular) grid that has true elevation values.
+
+    Parameters
+    ----------
+    lat_grid : ndarray
+        A 2d array containing latitude values that correspond to the other
+        two input grids.
+
+    lon_grid : ndarray
+        A 2d array containing longitude values that correspond to the other
+        two input grids.
+
+    elev_grid : ndarray
+        A 2d array containing elevation values that correspond to the other
+        two input grids.
+
+    num_samples : tuple
+        A tuple containing two integers. The first is the desired number of
+        points along the radial axis of the polar grid. The second is the
+        desired number of points along the angular axis of the polar grid.
+
+
+    Returns
+    -------
+    all_samples: list
+       Array of [lat, lon, elev] points that were sampled using the polar grid.
+
+    """
+
+    lats = lat_grid[0]
+    lons = lon_grid.T[0]
+
+    lat_range = lats[-1] - lats[0]
+
+    grid_shape = lat_grid.shape
+    grid_center_i = (grid_shape[0] - 1) // 2
+    grid_center_j = (grid_shape[1] - 1) // 2
+    center_lat = lat_grid[grid_center_i, grid_center_j]
+    center_lon = lon_grid[grid_center_i, grid_center_j]
+
+    try:
+        from scipy.interpolate import RegularGridInterpolator
+    except ImportError:
+        raise ImportError('The polar sampling function requires scipy')
+
+    interpolator = RegularGridInterpolator((lats, lons), elev_grid.T)
+
+    r = np.linspace(0, lat_range//2, num_samples[0])
+    theta = np.linspace(0, 2 * np.pi, num_samples[1])
+    polar_pts = np.array(list(itertools.product(r, theta)))
+
+    pts = np.array([tools.polar_to_cart(e[0], e[1]) for e in polar_pts])
+    pts += np.array((center_lat, center_lon))
+    total_num_samples = num_samples[0]*num_samples[1]
+
+    interpolated_elevs = interpolator(pts).reshape(total_num_samples, 1)
+    samples = np.hstack((pts, interpolated_elevs))
+    return samples
+
+
+def uniformly_sample_triangle(p1, p2, p3, num_samples):
```

Comment:
What's the purpose of this public method?

Reply:
will make private